### ISSUE TITLE
IntelliSense process should not fail due to invalid characters in a MACRO name (i.e. ":")

### ISSUE THREAD

**Issue by dcourtois** [microsoft/vscode-cpptools#4028]
- OS and Version: Windows 10
- VS Code Version: 1.36.1
- C/C++ Extension Version: 0.24.1

As the title says, I have a weird issue. A few days ago, Intellisense started behaving really badly, and ultimately stopped working on some of my projects. On the same setup, some other projects work perfectly, so it might be related to the complexity of the CMake structure ?

Anyway, in the C/C++ output, I just have the following error: `Failed to read response from server: 22`. And when I check the running processes related to VSCode, I noticed that `Microsoft.VSCode.CPP.Extension.exe` is running, but there is no `Microsoft.VSCode.Cpp.Intellisense.MSVC.exe` process on the projects where I have the problem. On the ones where intellisense works, this process is running as usual.

In my settings, I'm using

```
	"C_Cpp.default.intelliSenseMode": "msvc-x64",
	"C_Cpp.default.configurationProvider": "vector-of-bool.cmake-tools",
```

But as I said, those settings are common to all my projects, and only a few of them exhibit the problem so it might not be related.

Problem is that I have no idea what to do to diagnose where this error might come from ... I tried uninstalling / reinstalling the extension, no luck. Is there anything I can do to find where this problem comes from ?

**Comment by michelleangela** [comment]
`Microsoft.VSCode.CPP.Extension.exe` is the process of the extension and `Microsoft.VSCode.Cpp.Intellisense.MSVC.exe` is the process for the IntelliSense server. The extension process creates one IntelliSense server process per translation unit (or source file that's currently open).

The error `Failed to read response from server: 22` indicates the extension process failed to read a response from the IntelliSense server after the extension sent an "update IntelliSense" request to the IntelliSense server.

To get more information on the data/requests the extension sends to the IntelliSense server, you can [set logging to debug level](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server).  Can you provide us these logs for the source file that is failing to get a response from the IntelliSense server? You can remove/edit any personal path files in the logs before posting it.

**Comment by dcourtois** [comment]
Hi,

First of all, thanks for the quick answer and the tip about logging. I'll add that to all my projects.

But for some reasons today things are back to working again... I did try a lot of different things yesterday before opening this issue, so maybe something worked... now what exactly, I'm not sure (I did uninstall/reinstall the C/C++ extension on VSCode, I uninstalled/reinstalled the Intellicode extension on Visual Studio because I thought the problems started to happen around the same time I started trying it, but that's probably unrelated, I wiped all my intellisense caches, etc.)

Anyway I'll leave the settings there, close this issue for now, and if intellisense break again, I'll reopen it and post the logs.

Thanks again !

**Comment by dcourtois** [comment]
Ok, I think I know what the problem was ! I've started using MSVC's `/DEBUG:FASTLINK` option to speed up my projects' link times. In my CMake file it's set like this `target_compile_options(mytarget PRIVATE $<$<CONFIG:Debug>:/DEBUG:FASTLINK>)` and it compiles fine (the option correctly appears in the `build.ninja` file generated by CMake.

But here is the log of the C/C++ extension:

```
Custom configurations received:
  uri: file:///d%3A/my_file.cc
  config: {
  "defines": [
    "NOMINMAX",
    "WIN32",
    "_WINDOWS",
    "NDEBUG",
    "EBUG:FASTLINK",
    "NDEBUG",
    "_CRT_SECURE_NO_WARNINGS"
  ],
  "standard": "c++17",
  "includePath": [
    // the list of paths is correct
  ],
  "intelliSenseMode": "msvc-x64",
  "compilerPath": "C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.16.27023/bin/Hostx64/x64/cl.exe"
}
textDocument/didOpen
textDocument/codeAction: 3
cpptools/activeDocumentChange
cpptools/textEditorSelectionChange
cpptools/textEditorSelectionChange
textDocument/documentSymbol: 4
cpptools/didChangeCustomConfiguration
Checking for syntax errors: file:///d%3A/my_file.cc
textDocument/documentSymbol: 5
sending compilation args for MY_FILE.CC
  include: C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\COMMUNITY\VC\TOOLS\MSVC\14.22.27905\INCLUDE
  include: C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\COMMUNITY\VC\TOOLS\MSVC\14.22.27905\ATLMFC\INCLUDE
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.18362.0\UM
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.18362.0\UCRT
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.18362.0\SHARED
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.18362.0\WINRT
  include: C:\PROGRAM FILES (X86)\WINDOWS KITS\10\INCLUDE\10.0.18362.0\CPPWINRT
  include: ... some other include folders
  define: _DEBUG
  define: UNICODE
  define: _UNICODE
  stdver: ms_c++latest
  intelliSenseMode: msvc
idle loop: reparsing the active document
Checking for syntax errors: file:///d%3A/my_file.cc
Shutting down IntelliSense server: MY_FILE.CC
Closing the communication channel.
Failed to create IntelliSense client. Can't create intellisense client for MY_FILE.CC
sending compilation args for MY_FILE.CC
  include: ...
  // again, long list of include directories that are correct
  define: NOMINMAX
  define: WIN32
  define: _WINDOWS
  define: NDEBUG
  define: EBUG:FASTLINK
  define: NDEBUG
  define: _CRT_SECURE_NO_WARNINGS
  stdver: ms_c++latest
  intelliSenseMode: msvc
queue_update_intellisense for files in tu of: MY_FILE.CC
queue_update_intellisense for files in tu of: MY_FILE.CC
errorSquiggles count: 1
errorSquiggles count: 1
Update IntelliSense time (sec): 1.462
Database safe to open
textDocument/hover: 6
Quick info operation failed: FE: 'Compiler exited with error - No IL available'
using Tag Parser for quick info
```

Everything seems right except the aforementioned option which appears as `EBUG:FASTLINK`

As I see it, there's 2 problems:
- CMake Tools extension clearly has a bug and interpret the beginning or the `/DEBUG:FASTLINK` as a 'this is a preprocessor's definition' and so incorrectly sends `EBUG:FASTLINK` to the C/C++ extension via the configuration provider's API. This is not your problem and I should probably report this to the CMake Tools' maintainers. **Edit:** I noticed the problem happens also with Visual Studio, so it's probably not a problem with CMake Tools but more a problem with CMake server or something.
- It seems that sending an invalid definition to the C/C++ extension makes the intellisense server fail. Which should not be the case. Any invalid definition should just be ignored and logged, and the intellisense service should be able to run as usual. So I'm reopening this issue to let you guys investigate on that part :)

If you need any more information, don't hesitate to ask.

Regards,
Damien.

**Comment by bobbrow** [comment]
Isn't `/DEBUG:FASTLINK` a linker option, not a compiler option?  The compiler interprets anything starting with `/D` as a constant or macro.  I agree we shouldn't crash, but you might want to consider updating your CMake recipe.  Should you be using `target_link_options` instead?

**Comment by dcourtois** [comment]
@bobbrow You're totally right, it was a mistake.
As a side note, `target_link_options` is a recent addition that I couldn't use (it was added in 3.13, while my lowest CMake is 3.12 on Linux)

**Comment by Colengms** [comment]
Having the extension filter out compiler args that it thinks are invalid would seem to be problematic.  If we assert that we know what arguments are valid, we need to maintain a complete database of the several hundred args for gcc/clang and CL.  That becomes fragile for users who want to use a compiler based on gcc, which has additional args.  We would need to keep track of every variant out there, or we would inadvertently strip args, leading to incorrect compiler probes, etc.  We should pass the same arguments to the compiler that the user specified.

However, it would seem to be a bug that the IntelliSense process fails if a define containing a colon is specified.  Currently, doing so, I get the following error:

```
Quick info operation failed: FE: 'Compiler exited with error - No IL available'
```

Updating the title to reflect.

**Comment by elahehrashedi** [comment]
The error message has been changed since this issue was reported.
Now, if there is a macro name defined that includes an invalid character, this squiggle message will be shown on the top of the file:
`command-line error: invalid macro definition: DEBUG:FASTLINK C/C++`
This is a better indication to show the user where the error is coming from.
A good change would be to add the squiggle to the c_cpp_properties.json to squiggle the defines that that includes an invalid character. You can follow this issue from here: https://github.com/microsoft/vscode-cpptools/issues/7674

**Comment by dcourtois** [comment]
Sorry to bother, but `showing a squiggled error` would absolutely *not* close or fix this issue. Printing a meaningful error is fine and all, BUT the real problem here is that it *shouldn't be an error* in the first place, and intellisense should *NOT* stop working on such constructs.

Reason why:
CMake 3.12 and earlier doesn't have the command `target_link_options` to correctly add options like `/DEBUG:FASTLINK`. So any CMake project that was created prior to the CMake 3.13 release (and probably a lot more since then, because people who write CMake projects tend to like being backward compatible, which means not using new commands as soon as they are released) are setting those options using the `target_compile_options`. This is *not* a bug or a malformed CMake construct. It's the way it was done, it's perfectly valid, and probably still commonly used.

So basically, in CMake, setting `/DEBUG:FASTLINK` in `target_compile_definitions` is *not* an error. At the very best, it could be considered a warning, but *only* if the CMake version used is greater than 3.13. And even then, we sometimes don't have the possibility to edit our CMake files because third parties, etc.

So again, Intellisense should just ignore thing it doesn't understand and continue working, with *at the user's discretion* an option to report those as warning/error.

**Comment by sean-mcmanus** [comment]
It's an error when input into our parser, but maybe we could workaround it.

@elahehrashedi Can we filter out macros with ":" that are not in a `\"`?

**Comment by dcourtois** [comment]
Thanks for taking the time to check this again.

> /DEBUG:FASTLINK is not a compiler arg [...]

I'm totally aware of that. As I said, prior to version 3.13 CMake had no way to separate those flags from the compile ones, and using `target_compile_definitions` was the "official" way to set such flags. And as such there are potentially lots of CMake projects that still set linker arguments like this.

> We could potentially filter out invalid macro names, and otherwise allow the file to be processed with the remaining configuration info. Though, that would seem to hide the underlying issue.

The first sentence is exactly what *I think* cpptools should do, and the second one is why *I think* it should optionally warn the user :) This way cpptools doesn't fail whenever there's something wrong, it still continue to work while still notifying the users that something's wrong.

As for the `/DEBUG:FASTLINK` specific use-case, I'll check if this still happens with CMake Tools, and report this there if it does. But again, that should not prevent cpptools to graciously handle such malformed "defines" :)

**Comment by dcourtois** [comment]
Sorry I clicked on the wrong button and mistakenly closed the issue with my previous comment :/

**Comment by Colengms** [comment]
Hi @dcourtois .  

Sure, we can keep this issue open to track also not giving up on a file if there is a malformed macro name.  However, I think that is a symptom of other problems, and those causes are important to address.

Note that it's not possible to specify a malformed macro name in code, and it's not possible to pass a malformed macro name to a gcc/clang via command line.  If you tried to pass in, `-DBAD:DEFINE=TEST", you would get the equivalent of "#define BAD :DEFINE TEST".  MSVC appears to simply ignore invalid macro names passed via command line (which is presumably not what the user is expecting, having gone to the trouble of passing it in.).

When you use `target_link_options`, do you also specify `/link` as the immediately prior option?  If not, it is probably not accomplishing what you expect.  The `/link` arg must precede `/DEBUG:FASTLINK` to be properly passed to the linker through a compiler command line.  If that is what you are doing, then it is a bug in CMake Tools that it's trying to parse args intended for the linker as compiler args.  It's important to get this fixed in CMake Tools, as the configuration is valid and being corrupted in such a way that cannot easily be addressed by the user (as the configuration is valid).  There are other potential linker args that could cause other unwanted behavior when parsed as compiler args.

Or, if you're not passing `/link` prior to `/DEBUG:FASTLINK`, that would appear to be user error, and CMake Tools is handling it properly.

**Comment by xZise** [comment]
Hi,

I'm not sure if my experience is related to this issue. Otherwise I'd be happy to open a separate issue.

On my machine I get a squiggle in the first character of every opened file:
`Befehlszeilenfehler: Ungültige Makrodefinition: 30Z)"C/C++(992)`
It mentions that there is an invalid macro definition and looking at the `c_cpp_properties.json` I have one entry in the `defines` which uses a colon `"VERSION_TIMED=\"3c23958 (tmp, 2021-12-21 17:30Z)\"". It also matches with the `30Z)` part of the error.

As soon as I remove the colon or it isn't followed by a digit the error disappears. Unfortunately the error message is in German even though VS code is set to English but maybe the error message can help identify the issue?

**Comment by sean-mcmanus** [comment]
The @xZise the scenario is the same, but it sounds like the original issue might be fixed and you may be getting "expected" behavior -- how exactly are you getting your compiler to build with the VERSION_TIMED? If your compiler is able to handle it and we're not then it'd be a bug.

**Comment by xZise** [comment]
I'm using platformio which it builds fine and it only affects intellisense for me. If you want I can upload a minimum example.

**Comment by sean-mcmanus** [comment]
I need to know what exactly is being passed to the compiler to define the VERSION_TIMED and what OS/compiler is being used and what the code is that uses the define (e.g. assigning it to string), because passing `VERSION_TIMED="3c23958 (tmp, 2021-12-21 17:30Z)"` to gcc doesn't compile. There are complicated rules in regards to the quoting. I suspect you're not defining VERSION_TIMED correctly -- we have an issue tracking improving the handling at https://github.com/microsoft/vscode-cpptools/issues/6773 .

**Comment by sean-mcmanus** [comment]
Your "define" should be `"-DVERSION_TIMED=\\\"8b312e3 (master, 2021-12-22 11:38Z)\\\""`, i.e. the `\"` becomes `"` so it needs the extra `\\` so that `\` is used. Does that fix your issue?
![image](https://user-images.githubusercontent.com/19859882/147497184-46bc345e-2421-4036-827c-91c6b7a5b4a6.png)

**Comment by xZise** [comment]
Unfortunately this doesn't fix it.

Maybe my understanding of C(++) is not good enough, but I'm wondering where the issue is exactly. Just using gcc/g++ with a define using a colon does work (at least how I do it), so I'm wondering what you do mean with:

> [...]because passing `VERSION_TIMED="3c23958 (tmp, 2021-12-21 17:30Z)"` to gcc doesn't compile[...]

Did you mean, that I needed to escape the quotes, something like here: `gcc -o c.out -DDEF_B=\"00:00\" main.c` ?

**Comment by sean-mcmanus** [comment]
What do you see for the VERSION_TIMED definition when you run C/C++ Log Diagnostics -- you should see `VERSION_TIMED="8b312e3 (master, 2021-12-22 11:38Z)"`. If you're defining it in a json file, and you want `\"` you need to use `\\\"` because of how `\` are handled. If you have compileCommands or configurationProvider set then it's not getting the definition from the c_cpp_properties.json. If you're defining it in the "defines" section it should be `"VERSION_TIMED=\"8b312e3 (master, 2021-12-22 11:38Z)\""`...my previously mentioned `"-DVERSION_TIMED=\\\"8b312e3 (master, 2021-12-22 11:38Z)\\\""` only works when it's used in the "compilerArgs".

**Comment by xZise** [comment]
Hi,

running C/C++ Log Diagnostics indeed reveals two entries:

    Defines:
        [...]
        VERSION="2e964ce (master)"
        VERSION_TIMED="2e964ce (master, 2022-01-10 18
        11Z)"
        Foo=Bar
        [...]

It also appears to output something that looks like JSON at the beginning and there it is split but also using escaped quotes:

```json
{
    "name": "PlatformIO",
    "includePath": [...],
    "browse": {
        "limitSymbolsToIncludedHeaders": true,
        "path": [...]
    },
    "defines": [
        "PLATFORMIO=50204",
        "ARDUINO_NodeMCU_32S",
        "GIT_REV=2e964ce",
        "VERSION=\"2e964ce (master)\"",
        "VERSION_TIMED=\"2e964ce (master, 2022-01-10 18",
        "11Z)\"",
        [...]
    ],
    "cStandard": "c99",
    "cppStandard": "c++11",
    "compilerPath": "/home/fabian/.platformio/packages/toolchain-xtensa32/bin/xtensa-esp32-elf-gcc",
    "compilerArgs": [
        "-mlongcalls"
    ],
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": false,
    "mergeConfigurations": false
}
```

The define is inside the `c_cpp_properties.json` looks like you mentioned (single `\"`) and it is not using `compilerArgs`:

```jsonc
//
// !!! WARNING !!! AUTO-GENERATED FILE!
// PLEASE DO NOT MODIFY IT AND USE "platformio.ini":
// https://docs.platformio.org/page/projectconf/section_env_build.html#build-flags
//
{
    "configurations": [
        {
            "name": "PlatformIO",
            "includePath": [...],
            "browse": {
                "limitSymbolsToIncludedHeaders": true,
                "path": [...]
            },
            "defines": [
                "PLATFORMIO=50204",
                "ARDUINO_NodeMCU_32S",
                "GIT_REV=2e964ce",
                "VERSION=\"2e964ce (master)\"",
                "VERSION_TIMED=\"2e964ce (master, 2022-01-10 18:11Z)\"",
                [...]
                ""
            ],
            [...]
            "compilerArgs": [
                "-mlongcalls",
                ""
            ]
        }
    ],
    "version": 4
}
```

Maybe that can narrow the issue down?

**Comment by sean-mcmanus** [comment]
Yes, the 
```
    VERSION_TIMED="2e964ce (master, 2022-01-10 18
    11Z)"
```
looks correct, but I don't know how to repro that. When I use `"VERSION_TIMED=\"2e964ce (master, 2022-01-10 18:11Z)\"",` in the defines, I get a log diagnostics output that shows `VERSION_TIMED="2e964ce (master, 2022-01-10 18:11Z)"`. Any idea what the difference could be?

**Comment by xZise** [comment]
Okay I did debug the extension and I've found the location where it is splitting. It read the value fine from the JSON, as you can see in the watched variable. Now `util.envDelimiter` is actually `:` and so I'd guess that this is the reason it is split along that.

![Bildschirmfoto von 2022-01-11 01-00-25](https://user-images.githubusercontent.com/417478/148858196-e106092b-dc14-4e88-94e0-15921d4ca438.png)

Now unfortunately I have no idea what this code is doing and whether this works as intended or is a bug hopefully it makes more sense to you?

**Comment by Colengms** [comment]
@xZise Thanks for digging into this!

@sean-mcmanus It looks like `updateConfigurationStringArray` uses `resolveAndSplit`, which appears to be intended only for paths.  It looks like we need 2 versions of this function, one for paths and one for non-paths, and the calls in `updateServerOnFolderSettingsChange` need to be updated to refer to the appropriate ones?

**Comment by sean-mcmanus** [comment]
Ah, okay -- I repro it now. It repros on Linux but not Windows. The original bug report said the issue was on Windows. Well, it repros on Windows if ";" is used instead of ":'. Yeah, seems like a resolveAndSplit problem.

**Comment by Colengms** [comment]
Reopening.  I think the issue this was previously tracking was not addressed.

**Comment by sean-mcmanus** [comment]
I moved the issue with ";" on Windows and ":" on Linux/Mac in defines/compilerArgs that got fixed to https://github.com/microsoft/vscode-cpptools/issues/8671 .
